registry/tokentransport.go

```diff
@@ -29,13 +29,16 @@ func (t *TokenTransport) RoundTrip(req *http.Request) (*http.Response, error) {
 
 	authService, err := isTokenDemand(resp)
 	if err != nil {
+		resp.Body.Close()
```

Comment:
Can't you just use `defer resp.Body.Close()` right after `resp, err := t.Transport.RoundTrip(req)`, or on line 28, and then you only have to worry about closing the one time?

Reply:
About using `defer resp.Body.Close()` directly after receiving `resp` in line 25: using defer here would mean closing the body when returning from this function, but the caller probably still wants to use the `resp.Body` afterwards, so closing it when returning closes it too early. According to https://golang.org/pkg/net/http/#Response it's explicitly in the responsibility of the caller/user of the HTTP client to close the body.

When closing it on line 28, it means basically the same. When I got the rest of the code right, `isTokenDemand` checks if the first response is a 401. If it is not, then the response with the body is returned to the caller in line 37, basically being a regular successful request without requiring any authentication magic. This is also the reason for the second `resp.Body.Close()` that happens when we have to authenticate as the body is then pretty uninteresting.